### DISCUSSION TITLE
Wiring up abort controller on RTKQ

### DISCUSSION THREAD

**Discussion by checkDaResume** [reduxjs/redux-toolkit#2048] [Q&A] [answered]
Hi all,

I kindly will like to ask a question about how abort control may be wired up with RTKQ. I understand that both base query and endpoint query have a signal API which I could tap into. However, I don't know where I will instantiate an abort controller and how would likely fire the abort command. My scenario is such that, should there be no token supplied to my auth slice via state then the fetch should be cancelled when an attempt to load this component occurs. Since signal is available on both baseQuery and endpoint query I will appreciate if someone could kindly explain the use case scenario for both. A sample of my code is here:

```js
export const massAPI = createApi({
    reducerPath: 'massAPI',
    baseQuery: fetchBaseQuery({
        baseUrl: 'http://localhost:3000/data/',
        prepareHeaders: (header, {getState}) => {
            const token = getState().users.user.token
            if(token){
                header.set('Authorization', `Bearer ${token}`)
            }else{

            }
            header.set('Content-Type', 'application/json')
            return header
        },
       // signal: <- "Access to signal here"
    }),
    endpoints: (builder)=>({
        getData: builder.query({
            query: (date) => ({
                url: `detail/${date}`,
                method: 'GET',
               // signal: <-"Access to signal here"
            }),
        })



```

Thanks in advance. Cheers.

**Comment by phryneas** [comment, marked as answer]
`signal` is already passed in - and you should not tinker with that. You can call `.cancel()` on your submitted query, but not add a new signal overriding the internal one.

Realistically I think it's the wrong tool for what you are doing here tbh. Because at that point the request is pretty much already done.

If you want to instantly return an error when there is no token in state yet, then better [wrap your baseQuery](https://redux-toolkit.js.org/rtk-query/usage/customizing-queries#automatic-re-authorization-by-extending-fetchbasequery) and just return an early error there.

**Reply by checkDaResume** [reply]
Thanks @phryneas I appreciate that response. You are absolutely right. I came across the option of wrapping baseQuery but since I am learning this tool I wanted to use this as an opportunity as well to see how an abort signal will be wired up (I did find mentions to Abortcontroller object but I couldn't find any documentation where signal is actually implemented in RTKQ). What I mean is how do I setup the condition for which I call the `.cancel()`. The below is perhaps wrong but just to show how I think this might work I will assume something similar to below:

```js
    endpoints: (builder)=>({
        getData: builder.query({
            query: (date, foo) => ({
                url: `detail/${date}`,
                method: 'GET',
                signal: (signal, foo)=> ({
                       if(foo){
                             signal.cancel()
                        }
                  })
            }),
        })

``` 

The `.foo` variable is being passed from the component where the Query hook is called. I will still go ahead and use the wrapping suggestion you have given as a better solution but also wish to understand this concept as well.

Thanks

**Reply by phryneas** [reply]
The thing is **that signal is already hooked up to RTK Query internals**. There is already an internal AbortController. That is not something for you to pass in.

**Reply by checkDaResume** [reply]
Okay I get you @phryneas. However, how is that signal object which is connected to the RTKQ internal used? might be a better question. Cos I don't assume that applying `.cancel()` to an endpoint directly without a condition is useful :-).

**Reply by phryneas** [reply]
```js
const [updateUser] = api.endpoints.updateUser.useMutation()


const updating = updateUser("foo")
// oh no, I changed my mind!
updating.abort()
```

Honestly, it's an absolute edge case.

There is generally very little value in aborting a request once it has started - the server has received it anyways, it will also continue processing it.

**Reply by checkDaResume** [reply]
LOL @phryneas thanks so much. You are a very good teacher. I appreciate you patience explaining this concept and find the comment on the code 💯 . This is exactly what I was trying to get from it. I will go forward and use the concept of wrapping baseQuery. Thanks again.

**Reply by checkDaResume** [reply]
It seems I might have another reason why I was considering the `.abort()` idea. So its established that it is a better idea to wrap the baseQuery for reauthorization based on `401 or 403`. However, if the user is logged-out then the query endpoint still keep firing on refresh / remount of component, it doesn't seem obvious a way to handle that promise. In the past (i.e. before I moved to RTKQ), I would send fetch request and then should there be no token supplied (i.e. no user), I will abort the fetch. With the mutation endpoints (like you have shown above), it seems relatively straight forward to call the`.cancel()` around some logic. However, with query endpoints, it doesn't appear that the endpoint methods are exposed by the hook. Let me illustrate what I mean:

```js
      [updateUser, {data, isLoading, isSuccess, ...etc}] = useUpdateUserMutation() <---- I have access to 'updateUser'
      [getUser, {data, isLoading, isSuccess, ....etc}] = useGetUserQuery() <---- I do NOT have access to 'getUser'

```
Of course, the issue I am having may just be an outright weakness in the logic for my component but I just like to make sure I am having the right approach. At the moment the data fetch is continuously firing even when the user is logged-out.

**Reply by phryneas** [reply]
You can always write your endpoint to `provide` a `"unauthorized"` tag on a 403 response and invalidate that on login.

**Reply by checkDaResume** [reply]
Brilliant. Thanks for the tip.

**Comment by khaledebdelli** [comment]
@checkDaResume https://redux-toolkit.js.org/api/createAsyncThunk#canceling-while-running
